packages/backend-common/src/discovery/HostDiscovery.ts

```diff
@@ -0,0 +1,97 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Config } from '@backstage/config';
+import { PluginEndpointDiscovery } from './types';
+import { readHttpServerOptions } from '@backstage/backend-app-api';
+
+/**
+ * HostDiscovery is a basic PluginEndpointDiscovery implementation
+ * that assumes that all plugins are hosted in a single deployment.
+ *
+ * The deployment may be scaled horizontally, as long as the external URL
+ * is the same for all instances. However, internal URLs will always be
+ * resolved to the same host, so there won't be any balancing of internal traffic.
+ *
+ * @public
+ */
+
+export class HostDiscovery implements PluginEndpointDiscovery {
+  /**
+   * Creates a new HostDiscovery discovery instance by reading
+   * from the `backend` config section, specifically the `.baseUrl` for
+   * discovering the external URL, and the `.listen` and `.https` config
+   * for the internal one.
+   *
+   * Can be overridden by in config by providing a `backend.discovery` section with a `<pluginId>` key.
+   *
+   * The basePath defaults to `/api`, meaning the default full internal
+   * path for the `catalog` plugin will be `http://localhost:7007/api/catalog`.
+   */
+  static fromConfig(config: Config, options?: { basePath?: string }) {
+    const basePath = options?.basePath ?? '/api';
+    const externalBaseUrl = config.getString('backend.baseUrl');
+
+    const {
+      listen: { host: listenHost = '::', port: listenPort },
+    } = readHttpServerOptions(config.getConfig('backend'));
+    const protocol = config.has('backend.https') ? 'https' : 'http';
+
+    // Translate bind-all to localhost, and support IPv6
+    let host = listenHost;
+    if (host === '::' || host === '') {
+      // We use localhost instead of ::1, since IPv6-compatible systems should default
+      // to using IPv6 when they see localhost, but if the system doesn't support IPv6
+      // things will still work.
+      host = 'localhost';
+    } else if (host === '0.0.0.0') {
+      host = '127.0.0.1';
+    }
+    if (host.includes(':')) {
+      host = `[${host}]`;
+    }
+
+    const internalBaseUrl = `${protocol}://${host}:${listenPort}`;
+
+    return new HostDiscovery(
+      internalBaseUrl + basePath,
+      externalBaseUrl + basePath,
+      config.getOptionalConfig('backend.discovery'),
+    );
+  }
+
+  private constructor(
+    private readonly internalBaseUrl: string,
+    private readonly externalBaseUrl: string,
+    private readonly discoveryConfig: Config | undefined,
+  ) {}
+
+  async getBaseUrl(pluginId: string): Promise<string> {
+    if (this.discoveryConfig?.has(pluginId)) {
+      return this.discoveryConfig.getString(pluginId);
```

Comment:
I'm feeling that breaking this down by endpoints or so might be the way to go, rather than per plugin. Perhaps something like this?

```
discovery:
  endpoints:
    - target: https://internal.example.com/internal-catalog
      plugins: [catalog]
    - target: https://internal.example.com/secure/api/{pluginId}
      plugins: [auth, permissions]
    - target:
        internal: https://internal.example.com/search
        external: https://example.com/search
      plugins: [search]
```

All examples and throwing out some extra ideas, I don't think we should implement all of those capabilities right now but it's a couple of things for the future. What do you think?

Reply:
Addressed in 960c1cf

Reply:
I've slightly changed the pattern to be `{{pluginId}}` or `{{ pluginId }}` to align with [`urlPatternDiscovery.compile`](https://backstage.io/docs/reference/core-app-api.urlpatterndiscovery.compile)